In [2]:
import pandas as pd 
import numpy as np

In [3]:
df= pd.read_csv('/Users/jonbeshahmadzai/Downloads/immovlan_dataset.csv')
df = df.drop(columns=["Open fire", "Surface area of the plot of land"])

<class 'pandas.DataFrame'>
RangeIndex: 22037 entries, 0 to 22036
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Locality                22036 non-null  str    
 1   Type of property        22036 non-null  str    
 2   Subtype of property     22036 non-null  str    
 3   Price                   21636 non-null  float64
 4   Type of sale            22036 non-null  str    
 5   Number of rooms         21564 non-null  float64
 6   Living Area             20611 non-null  float64
 7   Fully equipped kitchen  7124 non-null   float64
 8   Furnished               13074 non-null  float64
 9   Terrace                 20702 non-null  float64
 10  Terrace Area            8639 non-null   float64
 11  Garden                  20307 non-null  float64
 12  Garden Area             5098 non-null   float64
 13  Surface of the land     11727 non-null  float64
 14  Number of facades       16906 non-null  float64
 

Dropping duplicates:

In [4]:
df= df.drop_duplicates()


,Locality,Type of property,Subtype of property,Price,Type of sale,Number of rooms,Living Area,Fully equipped kitchen,Furnished,Terrace,Terrace Area,Garden,Garden Area,Surface of the land,Number of facades,Swimming pool,State of the building
0,1000 Brussels,Apartment,Apartment,365000.0,for sale,2.0,NaN,NaN,1.0,1.0,NaN,0.0,NaN,NaN,2.0,NaN,NaN
1,3070 Kortenberg,Apartment,Apartment,349000.0,for sale,2.0,79.0,0.0,0.0,1.0,NaN,0.0,NaN,NaN,NaN,0.0,Normal
2,4140 Sprimont,House,Residence,175000.0,for sale,2.0,75.0,0.0,0.0,0.0,NaN,1.0,NaN,46.0,3.0,0.0,Normal
3,1325 Chaumont Gistoux,House,Villa,783045.0,for sale,3.0,335.0,NaN,NaN,NaN,NaN,1.0,NaN,4500.0,4.0,NaN,New
4,8890 Moorslede,House,Residence,389000.0,for sale,4.0,220.0,NaN,NaN,1.0,29.0,1.0,NaN,411.0,2.0,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
22032,2800 Mechelen,Apartment,Duplex,380000.0,for sale,5.0,199.0,NaN,0.0,1.0,NaN,0.0,NaN,NaN,2.0,NaN,To be renovated
22033,1932 Woluwe Saint Etienne,House,Residence,515000.0,for sale,4.0,180.0,NaN,0.0,1.0,NaN,1.0,NaN,147.0,2.0,NaN,NaN
22034,1495 Marbais,House,Residence,415000.0,for sale,4.0,287.0,0.0,0.0,1.0,NaN,1.0,815.0,1030.0,3.0,0.0,To be renovated
22035,6983 Ortho,House,Residence,95000.0,for sale,1.0,77.0,NaN,1.0,0.0,NaN,1.0,407.0,407.0,4.0,0.0,To be renovated


Removing empty spaces: 

In [5]:
df = df.apply(lambda x: x.str.strip() if x.dtype == "object" else x)


,Locality,Type of property,Subtype of property,Price,Type of sale,Number of rooms,Living Area,Fully equipped kitchen,Furnished,Terrace,Terrace Area,Garden,Garden Area,Surface of the land,Number of facades,Swimming pool,State of the building
0,1000 Brussels,Apartment,Apartment,365000.0,for sale,2.0,NaN,NaN,1.0,1.0,NaN,0.0,NaN,NaN,2.0,NaN,NaN
1,3070 Kortenberg,Apartment,Apartment,349000.0,for sale,2.0,79.0,0.0,0.0,1.0,NaN,0.0,NaN,NaN,NaN,0.0,Normal
2,4140 Sprimont,House,Residence,175000.0,for sale,2.0,75.0,0.0,0.0,0.0,NaN,1.0,NaN,46.0,3.0,0.0,Normal
3,1325 Chaumont Gistoux,House,Villa,783045.0,for sale,3.0,335.0,NaN,NaN,NaN,NaN,1.0,NaN,4500.0,4.0,NaN,New
4,8890 Moorslede,House,Residence,389000.0,for sale,4.0,220.0,NaN,NaN,1.0,29.0,1.0,NaN,411.0,2.0,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
22032,2800 Mechelen,Apartment,Duplex,380000.0,for sale,5.0,199.0,NaN,0.0,1.0,NaN,0.0,NaN,NaN,2.0,NaN,To be renovated
22033,1932 Woluwe Saint Etienne,House,Residence,515000.0,for sale,4.0,180.0,NaN,0.0,1.0,NaN,1.0,NaN,147.0,2.0,NaN,NaN
22034,1495 Marbais,House,Residence,415000.0,for sale,4.0,287.0,0.0,0.0,1.0,NaN,1.0,815.0,1030.0,3.0,0.0,To be renovated
22035,6983 Ortho,House,Residence,95000.0,for sale,1.0,77.0,NaN,1.0,0.0,NaN,1.0,407.0,407.0,4.0,0.0,To be renovated


Checking Prices for error and urealistic prices:

In [ ]:
#replacing unrealistic prices with None (the limit i use is 1000 euros after checking the data and putting the price in ascending order)
df.loc[df['Price'] <=1000, 'Price'] = np.nan


Checking Number of Rooms (Error) and Unrealistic Number of Rooms : Wohoooooo

In [7]:
df[df["Number of rooms"] == 0].shape

(120, 17)

In [8]:
df[df["Number of rooms"] == 0].head(10)

,Locality,Type of property,Subtype of property,Price,Type of sale,Number of rooms,Living Area,Fully equipped kitchen,Furnished,Terrace,Terrace Area,Garden,Garden Area,Surface of the land,Number of facades,Swimming pool,State of the building
69,6980 La Roche En Ardenne,Apartment,Apartment,NaN,for sale,0.0,47.0,NaN,NaN,0.0,NaN,0.0,NaN,NaN,NaN,NaN,Normal
122,8620 Nieuwpoort,Apartment,Studio,112000.0,for sale,0.0,26.0,NaN,0.0,1.0,3.0,0.0,NaN,NaN,2.0,NaN,Normal
147,1060 Sint Gillis,Apartment,Studio,195000.0,for sale,0.0,30.0,NaN,0.0,1.0,6.0,0.0,NaN,NaN,NaN,0.0,Excellent
513,6940 Durbuy,House,Residence,NaN,for sale,0.0,30.0,NaN,NaN,1.0,NaN,0.0,NaN,190.0,NaN,NaN,NaN
536,7140 Morlanwelz,House,Residence,154000.0,for sale,0.0,172.0,0.0,NaN,1.0,NaN,1.0,NaN,NaN,NaN,NaN,NaN
745,8660 De Panne,Apartment,Studio,120000.0,for sale,0.0,31.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Excellent
1254,1410 Waterloo,Apartment,Studio,225000.0,for sale,0.0,57.0,NaN,0.0,1.0,20.0,1.0,150.0,NaN,2.0,NaN,NaN
1607,1060 Sint Gillis,Apartment,Studio,160000.0,for sale,0.0,31.0,1.0,NaN,0.0,NaN,0.0,NaN,NaN,2.0,0.0,Normal
1689,4000 Liege,Apartment,Studio,69000.0,for sale,0.0,42.0,0.0,0.0,1.0,3.0,0.0,NaN,NaN,2.0,0.0,To renovate
2197,4360 Bergilers,House,Residence,475000.0,for sale,0.0,NaN,NaN,0.0,0.0,NaN,0.0,NaN,1343.0,NaN,NaN,To renovate


In [9]:
df[df["Number of rooms"] == 0]["Subtype of property"].value_counts()

Subtype of property
Studio            85
Apartment         18
Residence         11
Mixed Building     2
Master House       1
Triplex            1
Ground Floor       1
Penthouse          1
Name: count, dtype: int64

In [10]:
#now cleaning number of rooms with 0 for Apartment, Residence, mixed building, master house, triplex ,groundfloor and penthouse. If it’s really a studio, the subtype should say Studio. When a penthouse could have 0 roomsThat would mean a penthouse studio. 

df.loc[(df["Number of rooms"] == 0) & (df["Subtype of property"] != "Studio"), "Number of rooms"] = np.nan


Now checking fully equiped kitchen. so since its a binary value between 1 = equiped and 0 no equiped. we can fill the missing values if its not mentioned in the listing as 0 = no equiped 

In [11]:
df['Fully equipped kitchen'] = df['Fully equipped kitchen'].fillna(0)

Now we do the same thing for furnished: assuming if they didn't mention it , its not equiped

In [12]:
df['Furnished'] = df['Furnished'].fillna(0)

Now Checking Living Area for Errors and Unrealistic data

In [13]:
df.sort_values("Living Area", ascending=False)[["Subtype of property","Living Area","Price"]].head(10)

,Subtype of property,Living Area,Price
4408,Cottage,13289.0,495000.0
8769,Mixed Building,5000.0,8900000.0
1112,Residence,3500.0,1400000.0
7320,Mansion,3000.0,2450000.0
3941,Mansion,3000.0,1350000.0
3379,Apartment,2800.0,6200000.0
14887,Residence,2700.0,13400000.0
20319,Mixed Building,2700.0,1725000.0
11690,Villa,2668.0,1795000.0
9,Cottage,2500.0,NaN


In [ ]:
#cleaning unrealistic vlaues for living area putting a maximum and then a minimum 
# Replacing unrealistic values (<10 m² or >5000 m²) with NaN
# These values likely cameme from scraping errors or misreported data

df.loc[df['Living Area'] > 10000 , 'Living Area'] = np.nan

df.loc[df['Living Area'] < 10 , 'Living Area'] = np.nan



In [15]:
df["Living Area"].describe()

count    20200.000000
mean       172.935099
std        148.971022
min         10.000000
25%         97.000000
50%        142.000000
75%        200.000000
max       5000.000000
Name: Living Area, dtype: float64

Now we do the same for Terrace : assuming if they didn't mention it , there is no terrace : filling the missing values with 0 as its binary. 

In [16]:
df['Terrace'] = df['Terrace'].fillna(0)

Now checking Terrace Area for Errors and Unrealistic Data and also if there is no terrace, the area cannot exist.

so Setting the terrace area to missing when there is no terrace.

In [17]:
df.loc[df["Terrace"] == 0, "Terrace Area"] = np.nan

It looks okay but still is 2968 m2 a bit suspecious?????????? gotta ask 

In [18]:
df.sort_values("Terrace Area", ascending=False)[["Locality","Subtype of property","Terrace Area","Price"]].head(10)

,Locality,Subtype of property,Terrace Area,Price
2138,6950 Harsin,Residence,2968.0,349000.0
2697,8930 Menen,Residence,1424.0,469000.0
15477,1390 Grez Doiceau,Residence,1040.0,1750000.0
12314,7520 Templeuve,Cottage,1000.0,749000.0
12766,6140 Fontaine L Eveque,Villa,912.0,490000.0
7780,6031 Monceau Sur Sambre,Villa,912.0,490000.0
2335,1190 Vorst,Apartment,865.0,279000.0
13025,5530 Yvoir,Apartment,796.0,300000.0
5237,1630 Linkebeek,Residence,700.0,1400000.0
7781,7850 Petit Enghien,Villa,600.0,995000.0


Also filling the Garden for missing Values: if there is a missing value then we fill it up with 0 assuming no garden :

In [19]:
df['Garden']= df['Garden'].fillna(0)

Now Checking Garden Area For Errors and Unrealistic Data 

also gotta do the same for if garden = 0 then garden area = None 

In [20]:
df.loc[df["Garden"] == 0, "Garden Area"] = np.nan

In [21]:
df.sort_values("Garden Area", ascending=False)[["Locality","Subtype of property","Garden Area","Price"]].head(10)

,Locality,Subtype of property,Garden Area,Price
1565,1130 Haren,Residence,690330.0,1100000.0
12749,6900 Marche En Famenne,Residence,118000.0,1200000.0
10496,7322 Pommeroeul,Mansion,85000.0,1750000.0
20500,4970 Stavelot,Residence,79850.0,1185000.0
13710,7910 Frasnes Lez Anvaing,Residence,70000.0,NaN
8606,1310 La Hulpe,Villa,69000.0,2900000.0
4087,1380 Lasne,Villa,69000.0,2900000.0
4191,3000 Leuven,Villa,61000.0,1590000.0
13576,5310 Eghezee,Residence,50000.0,NaN
7273,2160 Wommelgem,Cottage,47000.0,1095000.0


In [22]:
df.sort_values("Garden Area", ascending=True)[["Locality","Subtype of property","Garden Area","Price"]].head(10)

,Locality,Subtype of property,Garden Area,Price
9658,1701 Itterbeek,Residence,1.0,850000.0
18661,1428 Lillois Witterzee,Residence,1.0,750000.0
14241,5377 Baillonville,Chalet,1.0,95000.0
7498,1800 Vilvoorde,Residence,1.0,349000.0
18610,1601 Ruisbroek,Residence,1.0,290000.0
9471,1070 Anderlecht,Loft,1.0,379000.0
16970,5650 Walcourt,Villa,1.0,499000.0
4336,1740 Ternat,Residence,1.0,410000.0
15080,3080 Vossem,Residence,1.0,575000.0
5348,1501 Buizingen,Residence,1.0,385000.0


In [23]:
# Cleaning Garden Area
# Two clearly unrealistic values were detected during inspection:
# 1 m² (impossible garden) or maybe ???? gotta check
# 690330 m² (69 hectares, likely scraping error)
# These values are replaced with NaN

df.loc[df["Garden Area"] < 2, "Garden Area"] = np.nan
df.loc[df["Garden Area"] > 500000, "Garden Area"] = np.nan

In [24]:
df["Garden Area"].describe()

count      4993.000000
mean        939.417384
std        3850.576718
min           2.000000
25%          92.000000
50%         250.000000
75%         714.000000
max      118000.000000
Name: Garden Area, dtype: float64

Now dealing with Total Surface of the land if it has urealistic and errored data 

In [25]:
df.sort_values("Surface of the land", ascending=False)[["Locality","Subtype of property","Surface of the land","Price"]].head(10)

,Locality,Subtype of property,Surface of the land,Price
14703,9690 Kluisbergen,Cottage,178367.0,3650000.0
12749,6900 Marche En Famenne,Residence,101895.0,1200000.0
2542,3580 Beringen,Residence,96317.0,1750000.0
10496,7322 Pommeroeul,Mansion,87264.0,1750000.0
12879,1380 Lasne,Villa,86000.0,1675000.0
13710,7910 Frasnes Lez Anvaing,Residence,80000.0,NaN
20500,4970 Stavelot,Residence,79975.0,1185000.0
3012,1730 Asse,Villa,75203.0,1995000.0
4087,1380 Lasne,Villa,70000.0,2900000.0
8606,1310 La Hulpe,Villa,70000.0,2900000.0


In [26]:
df.sort_values("Surface of the land", ascending=True)[["Locality","Subtype of property","Surface of the land","Price"]].head(50)

,Locality,Subtype of property,Surface of the land,Price
7005,3080 Tervuren,Villa,1.0,1264580.0
805,1970 Wezembeek Oppem,Villa,1.0,1264580.0
3444,4420 Saint Nicolas,Residence,1.0,169000.0
20802,1652 Alsemberg,Residence,1.0,199000.0
16816,7321 Blaton,Residence,1.0,49900.0
9015,4030 Grivegnee,Residence,1.0,175000.0
2298,6240 Farciennes,Residence,3.0,185000.0
2765,6200 Chatelineau,Residence,3.0,185000.0
10286,7100 La Louviere,Residence,4.0,80000.0
2054,5330 Assesse,Residence,6.0,350000.0


In [27]:
# Cleaning Surface of the land
# Values smaller than 10 m² are unrealistic for land area
# These are replaced with NaN

df.loc[df["Surface of the land"] < 10, "Surface of the land"] = np.nan

Now we have to Deal with Number of Facades and its Errors and unrealistic data 

In [28]:
df.sort_values('Number of facades', ascending= False)[['Locality', 'Number of facades', 'Subtype of property', 'Price']].head()

,Locality,Number of facades,Subtype of property,Price
1051,1000 Brussels,8.0,Residence,1100000.0
998,1490 Court Saint Etienne,4.0,Villa,600000.0
8695,4053 Embourg,4.0,Residence,545000.0
5035,1910 Kampenhout,4.0,Villa,675000.0
5770,4682 Heure Le Romain,4.0,Bungalow,185000.0


In [29]:
# Cleaning unrealistic number of facades
# In Belgian real estate properties have between 1 and 4 facades.
# Values above 4 are likely scraping errors.

df.loc[df["Number of facades"] > 4, "Number of facades"] = np.nan

In [30]:
df['Number of facades'].describe()

count    16611.000000
mean         2.843658
std          0.850009
min          1.000000
25%          2.000000
50%          3.000000
75%          4.000000
max          4.000000
Name: Number of facades, dtype: float64

Now we come to the swimming pool part: Here i will try to fill the value of None to 0 indicating there is no swimming pool if they didnt mention it :  In real estate datasets, missing usually means there is no pool because agents almost always advertise pools.


In [31]:
df['Swimming pool']= df['Swimming pool'].fillna(0)

now we come to check the state of building : I can fill it with Unknow the missing values but for now i'll just leave it as its:

In [32]:
missing_values = df.isna().sum()

In [33]:
missing_values= df.isna().sum().sort_values(ascending=False)
missing_values

Garden Area               16628
Terrace Area              13177
Surface of the land       10076
Number of facades          5010
State of the building      4843
Living Area                1421
Number of rooms             500
Price                       389
Locality                      1
Type of property              1
Type of sale                  1
Subtype of property           1
Fully equipped kitchen        0
Furnished                     0
Terrace                       0
Garden                        0
Swimming pool                 0
dtype: int64

In [34]:
#Percentage of missing values per column?
missing_values = (df.isna().sum() / len(df)) * 100
missing_values.sort_values(ascending=False)


Garden Area               76.906711
Terrace Area              60.945377
Surface of the land       46.602840
Number of facades         23.171916
State of the building     22.399519
Living Area                6.572314
Number of rooms            2.312566
Price                      1.799177
Locality                   0.004625
Type of property           0.004625
Type of sale               0.004625
Subtype of property        0.004625
Fully equipped kitchen     0.000000
Furnished                  0.000000
Terrace                    0.000000
Garden                     0.000000
Swimming pool              0.000000
dtype: float64

Several columns contained a very small number of missing values, including Price (389), Locality (1), Type of property (1), Type of sale (1), and Subtype of property (1). Since these variables are essential for the analysis—particularly price, which is the primary variable used for evaluating and comparing properties—rows with missing values in these fields cannot provide meaningful information. Because the proportion of missing values for these columns represents only a very small percentage of the dataset, the affected rows were removed rather than attempting to impute values. This approach avoids introducing artificial or potentially incorrect information while maintaining the overall integrity and reliability of the dataset for further analysis.

In [35]:
df = df.dropna(subset=[
    "Price",
    "Locality",
    "Type of property",
    "Type of sale",
    "Subtype of property"
])

df = df.reset_index(drop=True)

In [36]:
#How are variables correlated to each other? (Why?)
df.select_dtypes(include='number').corr()

,Price,Number of rooms,Living Area,Fully equipped kitchen,Furnished,Terrace,Terrace Area,Garden,Garden Area,Surface of the land,Number of facades,Swimming pool
Price,1.000000,0.080770,0.626613,-0.037663,0.005813,0.082999,0.207933,0.097245,0.251174,0.293762,0.168139,0.186082
Number of rooms,0.080770,1.000000,0.720809,0.000676,-0.001939,-0.008851,0.151169,0.025603,0.133681,0.184984,0.036112,0.014331
Living Area,0.626613,0.720809,1.000000,0.002560,-0.029679,0.023055,0.270551,0.233547,0.292594,0.316634,0.249774,0.197421
Fully equipped kitchen,-0.037663,0.000676,0.002560,1.000000,0.017453,0.099526,0.025785,0.033017,-0.002590,-0.003820,-0.020167,0.009956
Furnished,0.005813,-0.001939,-0.029679,0.017453,1.000000,0.026176,0.009367,-0.058923,0.020356,0.035637,0.005070,0.056968
Terrace,0.082999,-0.008851,0.023055,0.099526,0.026176,1.000000,NaN,0.112408,0.013912,0.021183,-0.020078,0.079225
Terrace Area,0.207933,0.151169,0.270551,0.025785,0.009367,NaN,1.000000,0.131335,0.126880,0.134035,0.198690,0.177959
Garden,0.097245,0.025603,0.233547,0.033017,-0.058923,0.112408,0.131335,1.000000,NaN,0.057284,0.317489,0.120573
Garden Area,0.251174,0.133681,0.292594,-0.002590,0.020356,0.013912,0.126880,NaN,1.000000,0.778606,0.222235,0.171535
Surface of the land,0.293762,0.184984,0.316634,-0.003820,0.035637,0.021183,0.134035,0.057284,0.778606,1.000000,0.221488,0.144802


In [37]:
#What is the correlation between the variables and the price? (Why might that be?)
df.select_dtypes(include='number').corr()['Price'].sort_values(ascending=False)


Price                     1.000000
Living Area               0.626613
Surface of the land       0.293762
Garden Area               0.251174
Terrace Area              0.207933
Swimming pool             0.186082
Number of facades         0.168139
Garden                    0.097245
Terrace                   0.082999
Number of rooms           0.080770
Furnished                 0.005813
Fully equipped kitchen   -0.037663
Name: Price, dtype: float64

In [38]:
#Which variables have the greatest influence on the price?

In [39]:
#Which variables have the least influence on the price?

In [40]:
round(df.describe(),2)

,Price,Number of rooms,Living Area,Fully equipped kitchen,Furnished,Terrace,Terrace Area,Garden,Garden Area,Surface of the land,Number of facades,Swimming pool
count,21232.00,20754.00,19846.00,21232.00,21232.00,21232.00,8292.00,21232.00,4905.00,11319.00,16353.00,21232.00
mean,423686.26,3.15,171.93,0.13,0.03,0.73,27.31,0.59,915.76,1271.49,2.84,0.03
std,456424.56,13.75,146.76,0.34,0.17,0.44,52.80,0.49,3687.63,4202.76,0.85,0.18
min,1100.00,0.00,10.00,0.00,0.00,0.00,1.00,0.00,2.00,10.00,1.00,0.00
25%,225000.00,2.00,97.00,0.00,0.00,0.00,10.00,0.00,90.00,240.00,2.00,0.00
50%,340000.00,3.00,142.00,0.00,0.00,1.00,18.00,1.00,250.00,508.00,3.00,0.00
75%,495000.00,4.00,200.00,0.00,0.00,1.00,30.00,1.00,700.00,1055.50,4.00,0.00
max,26000000.00,1962.00,5000.00,1.00,1.00,1.00,2968.00,1.00,118000.00,178367.00,4.00,1.00
